src/auth/src/credentials/impersonated.rs

```diff
@@ -428,6 +428,60 @@ struct GenerateAccessTokenRequest {
     lifetime: String,
 }
 
+pub(crate) async fn generate_access_token(
+    source_headers: HeaderMap,
+    delegates: Option<Vec<String>>,
+    scopes: Vec<String>,
+    lifetime: Duration,
+    service_account_impersonation_url: &str,
+) -> Result<Token> {
+    let client = Client::new();
+    let body = GenerateAccessTokenRequest {
+        delegates,
+        scope: scopes,
+        lifetime: format!("{}s", lifetime.as_secs_f64()),
+    };
+
+    let response = client
+        .post(service_account_impersonation_url)
+        .header("Content-Type", "application/json")
+        .headers(source_headers)
+        .json(&body)
+        .send()
+        .await
+        .map_err(|e| errors::from_http_error(e, MSG))?;
+
+    if !response.status().is_success() {
+        let err = errors::from_http_response(response, MSG).await;
+        return Err(err);
+    }
+
+    let token_response = response
+        .json::<GenerateAccessTokenResponse>()
+        .await
+        .map_err(|e| {
+            let retryable = !e.is_decode();
+            CredentialsError::from_source(retryable, e)
+        })?;
+
+    let parsed_dt = OffsetDateTime::parse(
+        &token_response.expire_time,
+        &time::format_description::well_known::Rfc3339,
+    )
+    .map_err(errors::non_retryable)?;
+
+    let remaining_duration = parsed_dt - OffsetDateTime::now_utc();
+    let expires_at = Instant::now() + remaining_duration.try_into().unwrap();
```

Comment:
nit: (I see this code is just moved, but) I think we prefer `.expect("reason why this can't fail")` over `.unwrap()`

Reply:
I would end up writing something like expect("Remaining Duration should have been convertible into Instant but it failed" which is inferrable.